src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs

```diff
@@ -0,0 +1,573 @@
+                        return false;
+                    }
+
+                    if (context.Current == CloseBrace)
+                    {
+                        // This is an 'escaped' brace in a literal, like "{{foo" - keep going.
+                    }
+                    else
+                    {
+                        // This is an unbalanced close-brace, which is not allowed
+                        context.Error = Resources.TemplateRoute_MismatchedParameter;
+                        return false;
+                    }
+                }
+
+                if (!context.MoveNext())
+                {
+                    break;
+                }
+            }
+
+            var encoded = context.Capture();
+            var decoded = encoded.Replace("}}", "}").Replace("{{", "{");
+            if (IsValidLiteral(context, decoded))
+            {
+                parts.Add(RoutePatternFactory.LiteralPart(decoded));
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private static bool IsAllValid(TemplateParserContext context, List<RoutePatternPathSegment> segments)
+        {
+            // A catch-all parameter must be the last part of the last segment
+            for (var i = 0; i < segments.Count; i++)
+            {
+                var segment = segments[i];
+                for (var j = 0; j < segment.Parts.Count; j++)
+                {
+                    var part = segment.Parts[j];
+                    if (part.IsParameter &&
+                        ((RoutePatternParameterPart)part).IsCatchAll &&
+                        (i != segments.Count - 1 || j != segment.Parts.Count - 1))
+                    {
+                        context.Error = Resources.TemplateRoute_CatchAllMustBeLast;
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static bool IsSegmentValid(TemplateParserContext context, List<RoutePatternPart> parts)
+        {
+            // If a segment has multiple parts, then it can't contain a catch all.
+            for (var i = 0; i < parts.Count; i++)
+            {
+                var part = parts[i];
+                if (part.IsParameter && ((RoutePatternParameterPart)part).IsCatchAll && parts.Count > 1)
+                {
+                    context.Error = Resources.TemplateRoute_CannotHaveCatchAllInMultiSegment;
+                    return false;
+                }
+            }
+
+            // if a segment has multiple parts, then only the last one parameter can be optional 
+            // if it is following a optional seperator. 
+            for (var i = 0; i < parts.Count; i++)
+            {
+                var part = parts[i];
+
+                if (part.IsParameter && ((RoutePatternParameterPart)part).IsOptional && parts.Count > 1)
```

Comment:
Part is cast to `RoutePatternParameterPart` many times in the code below. Create a cast variable?

Reply:
This can probably be simplified now that we have pattern patching.